src/common/events/events_io.cpp

```diff
@@ -137,6 +137,76 @@ mio::ExpectedSize ReaderFromAsyncReader::Read(
 	return read;
 }
 
+TeeReader::TeeReaderLeafPtr TeeReader::AttachSecondaryReader() {
+	auto reader = make_shared<TeeReaderLeaf>(shared_from_this());
+	leaf_readers_.insert({reader, TeeReaderRequestedRead {}});
+	return reader;
+}
+
+error::Error TeeReader::ReadyToAsyncRead(
+	TeeReader::TeeReaderLeafPtr leaf_reader,
+	vector<uint8_t>::iterator start,
+	vector<uint8_t>::iterator end,
+	mio::AsyncIoHandler handler) {
+	// TODO: should this be an assert() instead?
+	auto found = leaf_readers_.find(leaf_reader);
+	if (found == leaf_readers_.end()) {
+		return error::MakeError(
+			error::ProgrammingError, "Leaf reader not found in internal map. This cannot happen");
+	}
+
+	leaf_readers_[leaf_reader].start = start;
+	leaf_readers_[leaf_reader].end = end;
+	leaf_readers_[leaf_reader].handler = handler;
+
+	if (++ready_to_read == leaf_readers_.size()) {
+		DoAsyncRead();
+		// TODO: should this be done from the _handler_ in DoAsyncRead()?
+		ready_to_read = 0;
+	}
+
+	return error::NoError;
+}
+
+void TeeReader::DoAsyncRead() {
+	auto handler = [this](mio::ExpectedSize result) {
+		if (!result) {
+			for (auto const &it : leaf_readers_) {
+				it.second.handler(result);
```

Comment:
We need to be very careful when calling handlers which can themselves modify `leaf_readers_`. It may be safer to make a copy of the list you're going to execute before you execute it.

Reply:
Right, thanks.